BuildMasterAutomation/Functions/ConvertFrom-BMOtterScriptExpression.ps1

```diff
@@ -0,0 +1,140 @@
+    a PowerShell representation of the object. This function supports converting both `vector` and `map` types into
+    their respective `array` and `hashtable` types in PowerShell.
+
+    .LINK
+    https://docs.inedo.com/docs/executionengine-otterscript-strings-and-literals
+
+    .EXAMPLE
+    "@(1, 2, 3, 4)" | ConvertFrom-BMOtterScriptExpression
+
+    Demonstrates converting an OtterScript vector into a PowerShell array.
+
+    .EXAMPLE
+    "%(hello: world)" | ConvertFrom-BMOtterScriptExpression
+
+    Demonstrates converting an OtterScript map into a PowerShell hashtable.
+
+    .EXAMPLE
+    "%(hello: %(hi: world))" | ConvertFrom-BMOtterSCriptExpression
+
+    Demonstrates converting nested OtterScript maps into nested PowerShell hashtables.
+
+    .EXAMPLE
+    "@(1, 2, @(3, 4))" | ConvertFrom-BMOtterScriptExpression
+
+    Demonstrates converting nested OtterScript vectors into nested PowerShell arrays.
+
+    .EXAMPLE
+    "@(1, 2, %(hello: world, hi there: @(5, 6, 7)))" | ConvertFrom-BMOtterScriptExpression
+
+    Demonstrates converting nested OtterScript vectors and maps into nested PowerShell arrays and hashtables.
+    #>
+    [CmdletBinding()]
+    param(
+        # The string to parse into a PowerShell expression
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [String] $Value,
+        [int] $Depth = 0
+    )
+
+    begin {
+        Set-StrictMode -Version 'Latest'
+        Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+
+        function ConvertTo-Int
+        {
+            param(
+                [Parameter(Mandatory)]
+                [String] $Current
+            )
+
+            $current = $Current.Trim()
+            $maybeInt = 0
+            if (([Int64]::TryParse($current, [ref] $maybeInt)))
+            {
+                return $maybeInt
+            }
+
+            return $current
+        }
+    }
+
+    process {
+        $isMap = $Value.StartsWith('%(') -and $Value.EndsWith(')')
+        $isVector = $Value.StartsWith('@(') -and $Value.EndsWith(')')
+
+        if (-not $isMap -and -not $isVector)
+        {
+            return $Value
+        }
+
+        $Value = $Value -replace '^@\(' -replace '^%\(' -replace '\)$'
+        $index = 0
+        $stack = [System.Collections.Stack]::New()
+        $nestedStore = @{}
+        while ($index -lt $Value.Length)
+        {
+            if (($Value[$index] -eq '@' -or $Value[$index] -eq '%')-and $Value[$index + 1] -eq '(')
+            {
+                $stack.Push($index)
+                $index++
+            }
+
+            if ($Value[$index] -ne ')')
+            {
+                $index++
+                continue
+            }
+
+            $lastItem = $stack.Pop()
+            if ($stack.Count -ne 0)
+            {
+                $index++
+                continue
+            }
+
+            $keyName = "PLACEHOLDER_$($nestedStore.Count)"

```

Comment:
We can avoid having to keep track of stack depth and placeholder values by doing this in two steps.

First simply parse all items in the vector/map by walking the string until you reach each comma `,` character and then return that substring. That code will look something like:

```powershell
$start = 0
$end = 0
$closesNeeded = 0

foreach ($i in 0..($Value.Length))
{
    $char = $Value[$i]

    # If $char is '(' and previous $char is '@' or '%'
        $closesNeeded++
        continue

    # if $char is ')' and $closesNeeded
        $closesNeeded--
        # if not the last character
            continue

    # if $closesNeeded greater than 0
        continue

    # if $char is anything other than ',' and not the last character
        continue

    $end = $i
    $lengthOfSubstring = $end - $start
    
    $Value.Substring($start, $lengthOfSubstring).Trim() | Write-Output

    $start = $i + 1 # Skip over the current comma char
}
```

Then once you have all the items, you can recursively pass them back into this function to resolve them into PowerShell objects, which will do the work for any deeper nested expressions.